docs/source/concepts.rst

```diff
@@ -0,0 +1,299 @@
+embedding vectors in its memory in the forward pass and updates the
+gradients in the backward.
+
+With sharding, there is an added communication cost: each GPU needs to
+ask the other GPUs for embedding vector lookup and communicate the
+gradients computed as well. This is usually referred to as all2all
+communication. In TorchRec, for input data on a given GPU, we determine
+where the embedding shard for each part of the data is located and send
+it to the target GPU. That target GPU then returns the embedding vectors
+back to the original GPU. In the backward pass, the gradients are sent
+back to the target GPU and the shards are updated accordingly with the
+optimizer.
+
+As described above, sharding requires us to communicate the input data
+and embedding lookups. TorchRec handles this in three main stages, we
+will refer to this as the sharded embedding module forward that is used
+in training and inference of a TorchRec model,
+
+-  Feature All to All/Input distribution (``input_dist``)
+
+   -  Communicate input data (in the form of a ``KeyedJaggedTensor``) to
+      the appropriate device containing relevant embedding table shard
+
+-  Embedding Lookup
+
+   -  Lookup embeddings with new input data formed after feature all to
+      all exchange
+
+-  Embedding All to All/Output Distribution (``output_dist``)
+
+   -  Communicate embedding lookup data back to the appropriate device
+      that asked for it (in accordance with the input data the device
+      received)
+
+-  The backward pass does the same operations but in reverse order.
+
+We show this below in the diagram,
+
+.. figure:: _static/img/torchrec_forward.png
+   :alt: Visualizing the forward pass including the input_dist, lookup, and output_dist of a sharded TorchRec module
+   :align: center
+
+   *Figure 2: Forward pass of a table wise sharded table including the input_dist, lookup, and output_dist of a sharded TorchRec module*
+
+**************************
+ DistributedModelParallel
+**************************
+
+All of the above culminates into the main entrypoint that TorchRec uses
+to shard and integrate the plan. At a high level,
+``DistributedModelParallel`` does,
+
+-  Initialize environment by setting up process groups and assigning
+   device type
+
+-  Uses default shaders if no shaders are provided, default includes
+   ``EmbeddingBagCollectionSharder``
+
+-  Takes in provided sharding plan, if none provided it generates one
+
+-  Creates sharded version of modules and replaces the original modules
+   with them, such as ``EmbeddingCollection`` to
+   ``ShardedEmbeddingCollection``
+
+-  By default, wraps the ``DistributedModelParallel`` with
+   ``DistributedDataParallel`` to make the module both model and data
+   parallel
+
+***********
+ Optimizer
+***********
+
+TorchRec modules provide a seamless API to fuse the backwards pass and
+optimize step in training, providing a significant optimization in
```

Comment:
```suggestion
optimize the step in training, providing a significant optimization in
```

Reply:
i think this one i meant to write "optimizer step" which makes more sense with the sentence